src/backend/distributed/commands/multi_copy.c

```diff
@@ -3509,6 +3555,51 @@ CopyGetPlacementConnection(ShardPlacement *placement, bool stopOnFailure)
 }
 
 
+static MultiConnection *
+GetConnectionIfExecutorPoolSizeReached(HTAB *connectionStateHash, char *nodeName, int
+									   nodePort)
+{
+	List *connectionStateList =
+		ConnectionStateListToNode(connectionStateHash, nodeName, nodePort);
+
+	if (list_length(connectionStateList) < MaxAdaptiveExecutorPoolSize)
```

Comment:
can't we have a single place to decide if the current session can open more connections? If we have a single place then we can easily control how many connections there are from the current session to a specific worker node, and we can have a single method which returns true or false based on that information. We can possibly store that information in SharedConnStatsHash, or create something new.

What I am suggesting is basically to always execute some method when we are getting a new connection so that we know how many we have at any time, and with that information we can easily make sure than we don't open more than MaxAdaptiveExecutorPoolSize connections to a specific node at any time.

Is this doable? If not, could you explain why?

Reply:
(I also have some suggestions for this PR but if the answer of above is yes then it will change probably)

Reply:
There are several questions here, so several answers:

- Connection throttling is a separate concept, and it'll be implemented with a follow-up PR (https://github.com/citusdata/citus/issues/3838) 

- >  If we have a single place then we can easily control how many connections there are from the current session to a specific worker node

That is already the case. If you go one line above, you'd see `GetConnectionIfPlacementAccessedInXact`. If you check out the implementation, you'd see that Citus keeps track of which placements (hence workers) are accessed within the current transaction/session.

We can go over those lower level internal data structures to count connections per session. But, I'd rather use `CopyConnectionState ` to find the connection on which there are less placements are assigned. The idea is to distribute the load equally over connections. The lower level APIs cannot (and should not) know which connections are used during the COPY execution.

> This would also honor the shared connection pool size.

Shared connection pool size is different concept at the moment. Moving session specific information to a shared memory is an overkill. When you access a shared memory, you have to serialize the access (e.g., access one at a time). That means that, everything slows down significantly.

 In the long term, we may implement the stand-alone/shared executor process(es), where backends gives tasks to a shared backend/executor, which handles all executions. But, that's far away at the moment. (There is an issue about this that Marco opened, but I couldn't find now)

Reply:
My interpretation of Talha's question is that we should have something like:
```c
if (!CanOpenCopyConnection(...))
{
   connection = GetLeastUtilisedCopyConnection(..);
}
```

where CanOpenCopyConnection looks at max_adaptive_executor_pool_size in this PR, and is expanded to check max_shared_pool_size in the next PR.

Reply:
>  where CanOpenCopyConnection looks at max_adaptive_executor_pool_size in this PR, and is expanded to check max_shared_pool_size in the next PR.


As far as I can see, that's not very easy to  achieve with our  current implementation of `max_shared_pool_size`. 

Because, we only consider `max_shared_pool_size` when we actually need to establish a new  connection. So, we don't have an API to learn if we can open new connections beforehand.

That's why, I followed the adaptive executor's approach for COPY as well. First, make sure that the executor pool size has room to open new connections. If so, go ahead for new connection establishment. Qualify the new connections as `OPTIONAL` vs `REQUIRED`. If it is optional, and the connection  is throttled, we use an existing connection. In COPY case, we  pick least  utilized connection, in the executor we have a more sophisticated way of putting the tasks  to a queue in the `WorkerPool` and pop dynamically when any connection is ready to execute a new task.

Here is the draft implementation for COPY to honor `max_shared_pool_size`: https://github.com/citusdata/citus/blob/ce1c6a5f4e151440a5d4f1ac91f3e0b2ac5e3660/src/backend/distributed/commands/multi_copy.c#L3559-L3614